airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+        return ""
+      # first, we check to see if we've been given a null slice, and if so, we change the query accordingly
+      if (stream_slice['first_day'] is None or stream_slice['last_day'] is None):
+        return f"{self.incremental_column} IS NULL"
+      return f"{self.incremental_column} >= to_timestamp('{stream_slice['first_day']}', 'YYYY-MM-DD') AND {self.incremental_column} <= to_timestamp('{stream_slice['last_day']}', 'YYYY-MM-DD')"
+    
+    def generate_primary_key_filter(self):
+      # Note that netsuite tables can have composite primary keys, and we've so far 
+      # seen keys with up to 3 columns. 
+      # To deal with this, we use generic syntax for keysort pagination with composite keys
+      # https://stackoverflow.com/questions/56719611/generic-sql-predicate-to-use-for-keyset-pagination-on-multiple-fields 
+      # given n keys, we should end up with (n1 > val1) OR (n1 = val1 AND n2 > val2) OR (n1 = val1 AND n2 = val2 AND n3 > val3) OR ...
+      filter_string = ""
+      for i, key in enumerate(self.primary_key_column):
+        last_value_seen = self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        filters =[ f"{key} > {last_value_seen}"]
+        for j in range(i):
+          previous_key = self.primary_key_column[j]
+          previous_value = self.primary_key_last_value_seen[previous_key] if previous_key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+          filters.append(f"{previous_key} = {previous_value}")
+        filter_string_for_step = "(" + " AND ".join(filters) + ")"
+        if filter_string != "":
+          filter_string = filter_string + " OR " + filter_string_for_step
+        else:
+          filter_string = filter_string_for_step
+      return f"({filter_string})"
+
+
+    def generate_primary_key_filter_deprecated(self):
+      primary_key_filter = ""
+      # primary keys can be composite!  To deal with this, we add filters for each column
+      # This is an edge case.  In the standard case where there's only a single primary key column, 
+      # this could should only add a single filter
+      for key in self.primary_key_column:
+        last_value_seen = self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        if primary_key_filter != "":
+          primary_key_filter = primary_key_filter + " AND "
+        if primary_key_filter == "" and self.has_composite_primary_key():
+          primary_key_filter = primary_key_filter + f"{key} = {last_value_seen}"
+        else:
+          primary_key_filter = primary_key_filter + f"{key} > {last_value_seen}"
+      # if it's a composite key, we add a second column to try to "upgrade" to the next column value
+      # if we don't do this, we don't properly make our way up the table
+      # NOTE:  This only works when there are two columns in the primary key
```

Comment:
are tables with 3 primary keys supported? If so, how?

Reply:
They are!  This is done in the current primary key filter function.  This comment was in a deprecated function that was dead code